options/posix/generic/lookup.cpp

```diff
@@ -0,0 +1,135 @@
+#include <mlibc/lookup.hpp>
+#include <mlibc/debug.hpp>
+#include <bits/ensure.h>
+
+#include <frg/string.hpp>
+#include <mlibc/allocator.hpp>
+#include <string.h>
+#include <errno.h>
+#include <arpa/inet.h>
+#include <unistd.h>
+
+namespace mlibc {
+
+static frg::string<MemoryAllocator> read_dns_name(char *buf, char *&it) {
+	frg::string<MemoryAllocator> res{getAllocator()};
+	while (true) {
+		char code = *it++;
+		if ((code & 0xC0) == 0xC0) {
+			// pointer
+			uint8_t offset = ((code & 0x3F) << 8) | *it++;
+			auto offset_it = buf + offset;
+			return res + read_dns_name(buf, offset_it);
+		} else if (!(code & 0xC0)) {
+			if (!code)
+				break;
+
+			for (int i = 0; i < code; i++)
+				res += (*it++);
+
+			if (*(it + 1))
+				res += '.';
+		} else {
+			break;
+		}
+	}
+
+	return res;
+}
+
+int __lookup_name(struct address *buf, const char *name) {
+	mlibc::infoLogger() << "__lookup_name() called!" << frg::endlog;
+	frg::string<MemoryAllocator> request{getAllocator()};
+
+	struct dns_header header;
+	header.identification = htons(123);
+	header.flags = htons(0x100);
+	header.no_q = htons(1);
+	header.no_ans = htons(0);
+	header.no_auths = htons(0);
+	header.no_additional = htons(0);
+
+	request.resize(sizeof(header));
+	memcpy(request.data(), &header, sizeof(header));
+
+	const char *end = name;
+	while (*end != '\0') {
+		end = strchrnul(name, '.');
+		size_t length = end - name;
+		frg::string<MemoryAllocator> substring{name, length, getAllocator()};
+		name += length + 1;
+		request += char(length);
+		request += substring;
+	}
+
+	request += char(0);
+	request += (1 >> 8);
+	request += (1 & 0xFF);
+	request += (1 >> 8);
+	request += (1 & 0xFF);
+
+	struct sockaddr_in sin = {};
+	sin.sin_family = AF_INET;
+	sin.sin_port = htons(53);
+	if (!inet_aton("8.8.8.8", &sin.sin_addr)) {
+		mlibc::infoLogger() << "__lookup_name(): inet_aton() failed!" << frg::endlog;
+		return -1;
+	}
+
+	int fd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (fd < 0) {
+		mlibc::infoLogger() << "_lookup_name(): socket() failed" << frg::endlog;
+		return -1;
+	}
+
+	size_t sent = sendto(fd, request.data(), request.size(), 0,
+			(struct sockaddr*)&sin, sizeof(sin));
+	if (sent != request.size()) {
+		mlibc::infoLogger() << "__lookup_name(): sendto() failed to send everything" << frg::endlog;
+		mlibc::infoLogger() << errno << frg::endlog;
+		return -1;
+	}
+
+	char response[256];
+	ssize_t rlen;
+	int count = 0;
+	while ((rlen = recvfrom(fd, response, 256, 0, NULL, NULL)) >= 0) {
```

Comment:
So, we can receive more than one packet?

Reply:
Hmmm, I'm actually not sure about that... This is the way musl had it but I can figure out the reason why they had it like that..